### PR TITLE
Fix bug that causes versions to be overwritten in model metadata

### PR DESCRIPTION
**Pull request by CoderHam**

### CODE REVIEW

File: src/core/tritonserver.cc [hunk 1]
```diff
@@ -1093,69 +1092,66 @@ TRITONSERVER_ServerModelMetadata(
...
+  metadata.AddMember("name", rapidjson::StringRef(model_name), allocator);
 
   rapidjson::Value versions(rapidjson::kArrayType);
   for (const auto v : ready_versions) {
     auto version_str = std::to_string(v);
-    versions.PushBack(
-        rapidjson::Value(version_str.c_str(), version_str.size()),
-        metadata.GetAllocator());
+    rapidjson::Value version_val(version_str.c_str(), allocator);
+    versions.PushBack(version_val, allocator);
```
Comment (GuanLuo) [inline src/core/tritonserver.cc:1105]: @deadeyegoodwin according to rapidjson's doc, it is incorrect to use temporary value in `PushBack()`, or you need to invoke the move schema explicitly via `Move()`. The change here is okay and it implicitly uses the move schema.

File: src/core/tritonserver.cc [hunk 2]
```diff
@@ -1093,69 +1092,66 @@ TRITONSERVER_ServerModelMetadata(
...
+  auto& allocator = metadata.GetAllocator();
   metadata.SetObject();
   // Just store string reference in JSON object since it will be serialized to
   // another buffer.
-  metadata.AddMember(
-      "name", rapidjson::StringRef(model_name), metadata.GetAllocator());
+  metadata.AddMember("name", rapidjson::StringRef(model_name), allocator);
 
   rapidjson::Value versions(rapidjson::kArrayType);
   for (const auto v : ready_versions) {
```
Comment (GuanLuo) [inline src/core/tritonserver.cc:1102]: Don't you need to check and only return one version if `model_version` is specified?

**Comment by CoderHam** [comment]
> Looks good but I'm not seeing why this fixes anything. Is it because the rapidjson::Value constructor was taking a reference to the string instead of copying?

Yes. The reference was being over written each time the version_str was reassigned a new value and hence all of the `versions` in the `version` object pointed to the latest string version. If you used to allocator to create a copy of the string variable in a rapidjson::Value object it retains the original value.

File: src/core/tritonserver.cc (hunk 2 above)
Comment (CoderHam) [reply src/core/tritonserver.cc:1102]: Yes. I did not touch that logic but I can fix it in this PR itself if you think thats a good idea,

File: src/core/tritonserver.cc [hunk 3]
```diff
@@ -1093,69 +1092,72 @@ TRITONSERVER_ServerModelMetadata(
...
+  metadata.AddMember("name", rapidjson::StringRef(model_name), allocator);
 
   rapidjson::Value versions(rapidjson::kArrayType);
-  for (const auto v : ready_versions) {
-    auto version_str = std::to_string(v);
-    versions.PushBack(
-        rapidjson::Value(version_str.c_str(), version_str.size()),
-        metadata.GetAllocator());
+  if (model_version_int != -1) {
+    auto version_str = std::to_string(model_version_int);
```
Comment (GuanLuo) [inline src/core/tritonserver.cc:1103]: Need to check if it is one of the ready version...

Comment (GuanLuo) [reply src/core/tritonserver.cc:1103]: My bad, I think `GetInferenceBackend()` will return error in that case. The version should be ready if control reaches here.

**Pull request merged**